pkg/sql/pgwire/conn.go

```diff
@@ -430,37 +426,223 @@ Loop:
+//   and to ultimately declare the authentication successful.
+// reserved: Reserved memory. This method takes ownership.
+// cancelConn: A function to be called when this goroutine exits. Its goal is to
+//   cancel the connection's context, thus stopping the connection's goroutine.
+//   The returned channel is also closed before this goroutine dies, but the
+//   connection's goroutine is not expected to be reading from that channel
+//   (instead, it's expected to always be monitoring the network connection).
+func (c *conn) processCommandsAsync(
+	ctx context.Context,
+	authOpt authOptions,
+	ac AuthConn,
+	sqlServer *sql.Server,
+	reserved mon.BoundAccount,
+	cancelConn func(),
+) <-chan error {
+	retCh := make(chan error, 1)
+	go func() {
+		var retErr error
+		var connHandler sql.ConnectionHandler
+		var authOK bool
+		defer func() {
+			// Release resources, if we still own them.
+			if reserved != (mon.BoundAccount{}) {
+				reserved.Close(ctx)
+			}
+			// Notify the connection's goroutine that we're terminating. The
+			// connection might know already, as it might have triggered this
+			// goroutine's finish, but it also might be us that we're triggering the
+			// connection's death. This context cancelation serves to interrupt a
+			// network read on the connection's goroutine.
+			cancelConn()
+
+			pgwireKnobs := sqlServer.GetExecutorConfig().PGWireTestingKnobs
+			if pgwireKnobs != nil && pgwireKnobs.CatchPanics {
+				if r := recover(); r != nil {
+					// Catch the panic and return it to the client as an error.
+					pge := pgerror.NewErrorf(pgerror.CodeCrashShutdownError, "caught fatal error: %v", r)
+					pge.Detail = string(debug.Stack())
+					retErr = pge
+					_ = writeErr(
+						ctx, &sqlServer.GetExecutorConfig().Settings.SV, retErr,
+						&c.msgBuilder, &c.writerState.buf)
+					_ /* n */, _ /* err */ = c.writerState.buf.WriteTo(c.conn)
+					c.stmtBuf.Close()
+					// Send a ready for query to make sure the client can react.
+					// TODO(andrei, jordan): Why are we sending this exactly?
+					c.bufferReadyForQuery('I')
+				}
+			}
+			if !authOK {
+				ac.AuthFail(retErr)
+			}
+			// Inform the connection goroutine of success or failure.
+			retCh <- retErr
+		}()
+
+		// Authenticate the connection.
+		if !authOpt.skipAuth {
+			if authOpt.authHook != nil {
+				if retErr = authOpt.authHook(ctx); retErr != nil {
+					return
+				}
+			} else {
+				if retErr = c.handleAuthentication(
+					ctx, ac, authOpt.insecure, authOpt.ie, authOpt.auth,
+					sqlServer.GetExecutorConfig(),
+				); retErr != nil {
+					return
+				}
+			}
+		}
+
+		connHandler, retErr = c.sendInitialConnData(ctx, sqlServer)
+		if retErr != nil {
+			return
+		}
+		ac.AuthOK(connHandler)
+		authOK = true
+
+		// Now actually process commands.
+		retErr = sqlServer.ServeConn(ctx, connHandler, reserved, cancelConn)
+		reserved = mon.BoundAccount{} // we've passed ownership away
```

Comment:
This happens too late - if `ServeConn` panics, then `reserved` won't get set to empty, so the `defer` in this function will fire and close `reserved`, but the parent of `reserved` gets closed by the `defer` within `ServeConn`.

The fix here is not quite clear to me - maybe we should be passing `reserved` by pointer and checking its `mon` pointer, since closing the parent will nil out `mon`? That seems pretty gross too. Maybe the `defer` here has to recover and then rethrow? @andreimatei what do you think?